packages/swing-store/src/importer.js

```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
+      allMetadata.set(key, value);
     }
   }
 
-  // At this point we should have acquired the entire KV store state, plus
-  // sufficient metadata to identify the complete set of artifacts we'll need to
-  // fetch along with the information required to validate each of them after
-  // fetching.
-  //
-  // Depending on how the export was parameterized, the metadata may also include
-  // information about historical artifacts that we might or might not actually
-  // fetch depending on how this import was parameterized
-
-  // Fetch the set of current artifacts.
-
-  // Keep track of fetched artifacts in this set so we don't fetch them a second
-  // time if we are trying for historical artifacts also.
-  const fetchedArtifacts = new Set();
-
-  for await (const [vatID, vatInfo] of vatArtifacts.entries()) {
-    // For each vat, we *must* have a transcript span.  If this is not the very
-    // first transcript span in the history of that vat, then we also must have
-    // a snapshot for the state of the vat immediately prior to when the
-    // transcript span begins.
-    vatInfo.transcriptKey ||
-      Fail`missing current transcript key for vat ${q(vatID)}`;
-    const transcriptInfo = artifactMetadata.get(vatInfo.transcriptKey);
-    transcriptInfo || Fail`missing transcript metadata for vat ${q(vatID)}`;
-    let snapshotInfo;
-    if (vatInfo.snapshotKey) {
-      snapshotInfo = artifactMetadata.get(vatInfo.snapshotKey);
-      snapshotInfo || Fail`missing snapshot metadata for vat ${q(vatID)}`;
-    }
-    if (!snapshotInfo) {
-      transcriptInfo.startPos === 0 ||
-        Fail`missing current snapshot for vat ${q(vatID)}`;
+  // Now take each metadata record and install the stub/pruned entry
+  // into the DB.
+
+  for (const [key, value] of allMetadata.entries()) {
+    const [tag, ...pieces] = key.split('.');
+    if (tag === 'bundle') {
+      const bundleID = pieces[0];
+      internal.bundleStore.addBundleRecord(bundleID);
+    } else if (tag === 'snapshot') {
+      const [_vatID, endPos] = pieces;
+      if (endPos !== 'current') {
+        internal.snapStore.addSnapshotRecord(JSON.parse(value));
+      }
+      // metadata['snapshot.v1.current'] = 'snapshot.v1.5' , i.e. it
+      // points to the name of the current artifact. We could
+      // conceivably remember this and compare it against the .inUse
+      // property of that record, but it's not worth the effort (we
+      // might encounter the records in either order).
+    } else if (tag === 'transcript') {
+      const metadata = JSON.parse(value);
+      if (!!key.endsWith('.current') !== !!metadata.isCurrent) {
+        throw Fail`transcript key ${key} mismatches metadata ${metadata}`;
+      }
+      internal.transcriptStore.addTranscriptSpanRecord(metadata);
     } else {
-      snapshotInfo.snapPos + 1 === transcriptInfo.startPos ||
-        Fail`current transcript for vat ${q(vatID)} doesn't go with snapshot`;
```

Comment:
Yeah, I removed it. It worked in the old implementation because we gathered information about each vat as we walked the export data (the `vatArtifacts` Map of `vatInfo = { snapshotKey, transcriptKey }` records). That Map was then used to decide which vats to do anything with, and then populate both the metadata record *and* import the artifact for the current snapshot and/or transcript span. This is the approach that made it too easy to ignore the historical metadata.

In the new implementation, I don't build a table indexed by vats. I add all metadata, then I import all artifacts that the data source provides. So there's no point where I've got both a `snapshotInfo` and a `transcriptInfo` for the same vat.

(also, a weaker argument, I didn't want the import process memory usage to be O(N) in the number of vats, however my decision to de-duplicate/apply-deletions the exportData into `allMetadata` undermines that.. it's actually O(N) in the number of metadata records, which is far larger than the number of vats)

I thought it would be better to add this to the completeness check, than try to track it in the import loops. But the completeness checks are delegated to the individual stores, none of which get to see both transcripts and snapshots at the same time.

So I figured the check wasn't really that important, and too hard to implement, so I left it out. I did leave a TODO in `importer.js > assertComplete` which would do more than the original:

```js
export function assertComplete(internal, level) {
  ..
  internal.bundleStore.assertComplete(level);
  internal.transcriptStore.assertComplete(level);
  internal.snapStore.assertComplete(level);

  // TODO: every isCurrent span that starts with load-snapshot has a
  // matching snapshot (counter-argument: swing-store should not know
  // those details about transcript entries)
}
```